# Description: add aspect ratio estimates to resolution drop-down menu
# Upstream: http://bugzilla.gnome.org/show_bug.cgi?id=549642
#
diff -Nur -x '*.orig' -x '*~' gnome-control-center-2.27.4.1/capplets/display/xrandr-capplet.c gnome-control-center-2.27.4.1.new/capplets/display/xrandr-capplet.c
--- gnome-control-center-2.27.4.1/capplets/display/xrandr-capplet.c	2009-07-27 22:04:01.052549961 +0100
+++ gnome-control-center-2.27.4.1.new/capplets/display/xrandr-capplet.c	2009-07-27 22:04:03.184554067 +0100
@@ -437,6 +437,49 @@
 	combo_select (app->rotation_combo, _("Normal"));
 }
 
+static const char *
+generate_mode_text (int width, int height)
+{
+    int ratio;
+    const char *aspect = NULL;
+
+    if (width && height) {
+        if (width > height) ratio = width * 10 / height;
+        else ratio = height * 10 / width;
+
+        switch (ratio) {
+        case 13:
+            aspect = "4:3";
+            break;
+        case 16:
+            aspect = "16:10";
+            break;
+        case 17:
+            aspect = "16:9";
+            break;
+        case 12:
+            aspect = "5:4";
+            break;
+        // This catches 1.5625 as well (1600x1024) when maybe it shouldn't.
+        case 15:
+            aspect = "3:2";
+            break;
+        case 18:
+            aspect = "9:5";
+            break;
+        case 10:
+            aspect = "1:1";
+            break;
+        }
+    }
+
+    if (aspect)
+        return idle_free (g_strdup_printf (_("%d x %d (%s)"), width, height, aspect));
+    else
+        return idle_free (g_strdup_printf (_("%d x %d"), width, height));
+}
+
+
 static void
 rebuild_rate_combo (App *app)
 {
@@ -629,7 +672,7 @@
 	height = gnome_rr_mode_get_height (modes[i]);
 
 	add_key (app->resolution_combo,
-		 idle_free (g_strdup_printf (_("%d x %d"), width, height)),
+		 generate_mode_text (width, height),
 		 width, height, 0, -1);
 
 	if (width * height > best_w * best_h)
@@ -639,15 +682,13 @@
 	}
     }
 
-    current = idle_free (g_strdup_printf (_("%d x %d"),
-					  app->current_output->width,
-					  app->current_output->height));
+	current = generate_mode_text (app->current_output->width,
+					      		app->current_output->height);
 
     if (!combo_select (app->resolution_combo, current))
     {
 	combo_select (app->resolution_combo,
-		      idle_free (
-			  g_strdup_printf (_("%d x %d"), best_w, best_h)));
+			  generate_mode_text (best_w, best_h));
     }
 }
 
